src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -340,17 +410,16 @@ Status CoreWorkerDirectTaskSubmitter::CancelTask(TaskSpecification task_spec,
       return Status::OK();
     }
 
-    auto scheduled_tasks = task_queues_.find(scheduling_key);
+    auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+    auto &scheduled_tasks = scheduling_key_entry.task_queue;
     // This cancels tasks that have completed dependencies and are awaiting
     // a worker lease.
-    if (scheduled_tasks != task_queues_.end()) {
-      for (auto spec = scheduled_tasks->second.begin();
-           spec != scheduled_tasks->second.end(); spec++) {
+    if (!scheduled_tasks.empty()) {
+      for (auto spec = scheduled_tasks.begin(); spec != scheduled_tasks.end(); spec++) {
         if (spec->TaskId() == task_spec.TaskId()) {
-          scheduled_tasks->second.erase(spec);
+          scheduled_tasks.erase(spec);
 
-          if (scheduled_tasks->second.empty()) {
-            task_queues_.erase(scheduling_key);
+          if (scheduled_tasks.empty()) {
             CancelWorkerLeaseIfNeeded(scheduling_key);
```

Comment:
Do we need to check whether we should delete the SchedulingKeyEntry here?

Reply:
Yes! However, I think that the check should be placed inside the callback function in `CancelWorkerLeaseIfNeeded`, because the callback can call `CancelWorkerLeaseIfNeeded` as well, and `CancelWorkerLeaseIfNeeded` needs to access the SchedulingKeyEntry